Move the common test functions from cmd_test.go to cmd/testing/fake.go

asalkeld opened the pull request:
**What this PR does / why we need it**:
This is so that we can use NewAPIFactory() from cmd/set/*test.go
Up until now we would get a import loop error.

This commit also adds a basic unit test case for cmd/set/set_image.go

**Which issue this PR fixes**
partial #34592

**Special notes for your reviewer**:
None

**Release note**:

``` release-note
NONE
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34683)

<!-- Reviewable:end -->

deads2k commented:
This needs to hold for https://github.com/kubernetes/kubernetes/pull/33180

deads2k commented:
@fabianofranz I think this has implications downstream.  If the pattern works well, this could give you command line unit tests.

deads2k commented:
@soltysh fyi, since you've bugged me about them before. :)

fabianofranz commented:
Nice move.

asalkeld commented:
@fabianofranz PTAL

fabianofranz commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -0,0 +1,541 @@
+	"k8s.io/kubernetes/pkg/runtime/serializer"
+)
+
+type internalType struct {
```
Once you are here, should we make all these types exported? Like `InternalType`, `ExternalType`, and `ExternalType2`. Unless they are embedded in public structs, there isn't much sense in having unexported structs with exported fields.

fabianofranz commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -0,0 +1,541 @@
+
+var validVersion = registered.GroupOrDie(api.GroupName).GroupVersion.Version
+var internalGV = unversioned.GroupVersion{Group: "apitest", Version: runtime.APIVersionInternal}
+var UnlikelyGV = unversioned.GroupVersion{Group: "apitest", Version: "unlikelyversion"}
```
Same here, make all these vars exported?

fabianofranz commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -0,0 +1,541 @@
+	return scheme, mapper, codec
+}
+
+// TODO move to fake.go
```
This TODO can be removed?

fabianofranz commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -0,0 +1,541 @@
+
+// TODO move to fake.go
+
+type testFactory struct {
```
Same for this one?

fabianofranz reviewed (changes requested):
Minor nits about things that were here already, otherwise LGTM.

Commit:
Move the common test functions from cmd_test.go to cmd/testing/fake.go

This is so that we can use NewAPIFactory() from cmd/set/*test.go
Up until now we would get a import loop error.

This commit also adds a basic unit test case for cmd/set/set_image.go

asalkeld commented:
@fabianofranz I made the changes you suggested.

fabianofranz commented:
LGTM, @deads2k mind tagging?
